src/panic.md

```diff
@@ -0,0 +1,102 @@
+Rust frames, just as C++'s `throw` unwinds C++ frames, until the panic reaches
+the point of recovery (for instance at a thread boundary). This means that as
+the panic traverses Rust frames, live objects in those frames that [implement
+`Drop`][destructors] will have their `drop` methods called. Thus, when normal
+execution resumes, no-longer-accessible objects will have been "cleaned up"
+just as if they had gone out of scope normally.
+
+> **Note**: As long as this guarantee of resource-cleanup is preserved,
+> "unwinding" may be implemented without actually using the mechanism used by
+> C++ for the target platform.
+
+> **Note**: The standard library provides two mechanisms for recovering from a panic,
+> [`catch_unwind`][fn-catch-unwind] (which enables recovery within the
+> panicking thread) and [`thread::spawn`][thread-spawn] (which automatically
+> sets up panic recovery for the spawned thread so that other threads may
+> continue running).
+
+### Unwinding across FFI boundaries
+[unwinding-ffi]: #unwinding-across-ffi-boundaries
+
+It is possible to unwind across FFI boundaries using an [appropriate ABI
+declaration][unwind-abi]. While useful in certain cases, this creates unique
+opportunities for undefined behavior, especially when multiple language runtimes
+are involved.
+
+Unwinding with the wrong ABI is undefined behavior:
+
+* Causing an unwind into Rust code from a foreign function that was called via a
+  function declaration or pointer declared with a non-unwinding ABI, such as `"C"`,
+  `"system"`, etc. (For example, this case occurs when such a function written in
+  C++ throws an exception that is uncaught and propagates to Rust.)
+* Calling a Rust `extern` function that unwinds (with `extern "C-unwind"` or
+  another ABI that permits unwinding) from a runtime that does not support.
+  unwinding, such as code compiled with GCC or Clang using `-fno-exceptions`
+
+Catching a foreign unwinding operation (such as a C++ exception) using
+`catch_unwind`, `JoinHandle::join`, or by letting it propagate all the way to a
+Rust `main()` function will have one of two behaviors, and it is unspecified
+which will occur:
+* The process aborts.
+* The function returns a `Result::Err` containing an opaque type.
+
+Note that Rust code compiled or linked with a different runtime counts as a
+"foreign exception" for the purpose of this guarantee. Thus, a library that
+uses `panic!` and is linked against one version of the Rust standard library,
+invoked from an application that uses a different version of the standard
+library, may cause the entire application to crash even if the library is only
+used within a child thread.
+
+There are currently no guarantees about the behavior that occurs when a foreign
+runtime attempts to dispose of, or rethrow, a Rust `panic` payload. In other
+words, an unwind originated from a Rust runtime must either lead to termination
+of the process or be caught by the same runtime.
```

Comment:
Isn't "catch and ignore" also an option?

In particular, we guarantee that if we catch a panic from another Rust runtime, it will either abort or be returned as an opaque error type. But this paragraph seems to exclude the "opaque error type" possibility?

Reply:
That would fall under attempting to "dispose of" the payload, wouldn't it?

This was explicitly left undefined along with catching foreign exceptions, and when we established that catching foreign exceptions wouldn't be UB, we didn't make any corresponding decision about the reverse situation, other runtimes catching Rust panics.

The Rust-Rust case is somewhat unique in that, clearly, we've implied that there *is* a way to safely catch a Rust panic. But in the "two separate runtimes" case, I don't understand how we avoid the issues that led to this being UB in the first place.

Reply:
@workingjubilee @Amanieu @bjorn3, do any of you know what would be required for a foreign runtime to safely catch and discard or rethrow a Rust panic?

Reply:
So we are unilaterally saying that disposing of C++ exceptions is okay (or does C++ say that somewhere?), but disposing of Rust exceptions is not? That doesn't seem quite fair.^^

Reply:
Correctly disposing of a Rust panic in another language is currently not possible. It is not unsafe, but will cause any future panics to be treated as double panic and thus abort due to the panic count not being decremented as it should.

Reply:
When a foreign language catches and disposes of an exception from another runtime (Itanium EH), it's supposed to call the `unwind_cleanup` function with `_URC_FOREIGN_EXCEPTION_CAUGHT`. Rust could use that to do any cleanup it needs to.

It's generally unsound to rethrow a foreign exception without coordinating with the other runtime.

Reply:
Do we correctly call that function we we catch and dispose of a foreign exception?

Reply:
Looking at what we [currently do there](https://github.com/rust-lang/rust/blob/125b26acf6f3127ecdd344c372691cefe0e9243e/library/panic_unwind/src/gcc.rs#L74-L82),  we seem to end up [here](https://github.com/rust-lang/rust/blob/ec0e16a66571458dc5c296ccfa5e4d8bdcc2a2a8/library/std/src/panicking.rs#L56-L58):

```rust
extern "C" fn __rust_drop_panic() -> ! {
    rtabort!("Rust panics must be rethrown");
}
```
So, it's not UB, and there's no problem with the panic count because we abort the process.

The error isn't very nice though, since we also do not provide a way for a foreign runtime to rethrow the exception.

(I see a similar call in `seh.rs` so maybe the same happens on Windows?)

Reply:
Can we punt this to a separate issue+PR?

I think, ideally, we'd like to be a "good citizen" in the sense that other runtimes can safely catch and dispose of (or possibly even rethrow?) a Rust panic, and that the requirements for doing so correctly (for a given Rust implementation) are documented somewhere. But I think that's separate from making sure that the Reference correctly documents the guarantees we've already gotten approval for from the lang team.

Reply:
Yes, I agree and suggested the same above.

Can you open an issue?

Reply:
Looking over the possible types of `rust` issues, I'm not sure where this belongs; I know I've let the project-repo fall somewhat stagnant, but should I open an [issue there](https://github.com/rust-lang/project-ffi-unwind/issues)? Or should I open a blank issue in the `rust` repo?

I don't _think_ this warrants the full RFC process. But it does require that we:

* decide what guarantee Rust can make regarding disposing/rethrowing panics from other languages
* implement the changes to make that possible in the standard library
* document the guarantee
* document somewhere what our requirements are for code in other languages to handle panics

So perhaps this does warrant a short RFC?

Reply:
I'd make it a rust repo issue, yeah. We'd want input from... libs-api and maybe lang I guess?... on what should happen when a foreign language disposes a Rust panic:
- UB (not a great answer IMO)
- guaranteed abort
- abort or X, for some X

Rethrowing from other languages is a somewhat separate discussion IMO and less pressing, given that in general there is no portable way to rethrow any exception from a different runtime.

Reply:
https://github.com/rust-lang/rust/issues/130369